File: python/src/coding_interview/presentation/preparation.py
from __future__ import annotations

from decimal import Decimal, InvalidOperation

from coding_interview.domain.user_id import UserId
from coding_interview.presentation.exceptions import BadRequestException


def parse_user_id(s: str) -> UserId:
    try:
        return UserId(s)
    except (ValueError, Exception) as e:
        raise BadRequestException(str(e)) from e


def parse_amount(s: str) -> Decimal:
    try:
        return Decimal(s)
    except (InvalidOperation, Exception) as e:
        raise BadRequestException(f"invalid amount: {s}") from e

File: golang/internal/presentation/preparation.go
package presentation

import (
	"fmt"

	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

func parseUserId(s string) (domain.UserId, error) {
	uid, err := domain.NewUserId(s)
	if err != nil {
		return domain.UserId{}, newBadRequest(err.Error())
	}
	return uid, nil
}

func parseAmount(s string) (decimal.Decimal, error) {
	d, err := decimal.NewFromString(s)
	if err != nil {
		return decimal.Zero, newBadRequest(fmt.Sprintf("invalid amount: %s", s))
	}
	return d, nil
}
